src/librustc_typeck/check/mod.rs

```diff
@@ -4844,8 +4850,13 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                     }
                 }
                 GenericParamDefKind::Type {..} => {
-                    let (types, infer_types) = segment.map_or((&[][..], true), |(s, _)| {
-                        (s.parameters.as_ref().map_or(&[][..], |p| &p.types[..]), s.infer_types)
+                    let (types, infer_types) = segment.map_or((vec![], true), |(s, _)| {
+                        (s.args.as_ref().map_or(vec![], |data| {
+                            data.args.iter().filter_map(|arg| match arg {
+                                GenericArg::Type(ty) => Some(ty),
+                                _ => None,
+                            }).collect()
+                        }), s.infer_types)
```

Comment:
Oh, I see, this is where. Same thing I said for `astconv` applies here too - what's happening here is too inefficient, and will miss errors if we ever allow mixing lifetimes and types.